pkg/webhook/cspc_operations.go

```diff
@@ -509,3 +470,203 @@ func GetNewBDFromRaidGroups(newRG, oldRG *cstor.RaidGroup) map[string]string {
+		isRaidGroupExist := false
+		for _, newRg := range newPoolSpec.DataRaidGroups {
+			if IsRaidGroupCommon(oldRg, newRg) {
+				isRaidGroupExist = true
+				rgs["data"].oldRaidGroups = append(rgs["data"].oldRaidGroups, oldRg)
+				rgs["data"].newRaidGroups = append(rgs["data"].newRaidGroups, newRg)
+				break
+			}
+		}
+		// Old raid group should exist on new pool spec changes
+		if !isRaidGroupExist {
+			return nil, errors.Errorf("removing raid group from pool spec is invalid operation")
+		}
+	}
+	for _, oldRg := range oldPoolSpec.WriteCacheRaidGroups {
+		isRaidGroupExist := false
+		for _, newRg := range newPoolSpec.WriteCacheRaidGroups {
+			if IsRaidGroupCommon(oldRg, newRg) {
+				isRaidGroupExist = true
+				rgs["data"].oldRaidGroups = append(rgs["data"].oldRaidGroups, oldRg)
+				rgs["data"].newRaidGroups = append(rgs["data"].newRaidGroups, newRg)
+				break
+			}
+		}
+		// Old raid group should exist on new pool spec changes
+		if !isRaidGroupExist {
+			return nil, errors.Errorf("removing raid group from pool spec is invalid operation")
+		}
+	}
+	return rgs, nil
+}
+
+// ArePoolSpecChangesValid validates the pool specs on CSPC for raid groups
+// changes(day-2-operations). Steps performed in this function
+// 1. Get common raidgroups with index matching from old and new spec.
+// 2. Iterate over common old and new raid groups and perform following steps:
+//    2.1 Validate raid group changes.
+//        2.1.1: Verify and return error when new block device added or removed from existing
+//               raid groups for other than stripe pool type.
+//    2.2 Validate changes for blockdevice replacement scenarios(openebs/openebs#2846).
+// 3. Validate vertical pool expansions if there are any new raidgroups or blockdevices added.
+func (pOps *PoolOperations) ArePoolSpecChangesValid(oldPoolSpec, newPoolSpec *cstor.PoolSpec) (bool, string) {
+	newToOldBd := make(map[string]string)
+	commonRaidGroups, err := getIndexedCommonRaidGroups(oldPoolSpec, newPoolSpec)
+	if err != nil {
+		return false, fmt.Sprintf("raid group validation failed: %v", err)
+	}
+	for _, v := range commonRaidGroups {
+		rgType := v.rgType
+		for _, oldRg := range v.oldRaidGroups {
+			oldRg := oldRg
+			for _, newRg := range v.newRaidGroups {
+				newRg := newRg
+				if err = validateRaidGroupChanges(&oldRg, &newRg, rgType); err != nil {
+					return false, fmt.Sprintf("raid group validation failed: %v", err)
+				}
+				if IsBlockDeviceReplacementCase(&oldRg, &newRg) {
+					if ok, msg := pOps.IsBDReplacementValid(&newRg, &oldRg, rgType); !ok {
+						return false, msg
+					}
+					newBD := GetNewBDFromRaidGroups(&newRg, &oldRg)
+					for k, v := range newBD {
+						newToOldBd[k] = v
+					}
+				}
+				// Already mapped(via index) old raid groups and new raid groups in
+				// commonRaidGroups no need to check other raid groups
+				break
+			}
+		}
+		err = pOps.validatePoolExpansion(newPoolSpec, v, rgType)
```

Comment:
It should be out of the loop and pass the two arguments newPoolSpec & commonRaidGroups. Then we have to change the function also.
If we write out of the loop then we will not check again and again please look at func definition comment